packages/ic-pub-key/src/lib.rs

```diff
@@ -0,0 +1,415 @@
+            Ok(Self { inner })
+        }
+
+        #[cfg(not(feature = "secp256k1"))]
+        {
+            Err(Error::AlgorithmNotSupported)
+        }
+    }
+}
+
+impl TryFrom<&SchnorrKeyId> for MasterPublicKey {
+    type Error = Error;
+
+    fn try_from(key_id: &SchnorrKeyId) -> Result<Self, Self::Error> {
+        #[cfg(feature = "secp256k1")]
+        {
+            if key_id.algorithm == SchnorrAlgorithm::Bip340secp256k1 {
+                let key_id = match key_id.name.as_ref() {
+                    "key_1" => ic_secp256k1::MasterPublicKeyId::SchnorrKey1,
+                    "test_key_1" => ic_secp256k1::MasterPublicKeyId::SchnorrTestKey1,
+                    _ => return Err(Error::UnknownKeyIdentifier),
+                };
+
+                let mk = ic_secp256k1::PublicKey::mainnet_key(key_id);
+                let inner = MasterPublicKeyInner::Bip340Secp256k1(mk);
+                return Ok(Self { inner });
+            }
+        }
+
+        #[cfg(feature = "ed25519")]
+        {
+            if key_id.algorithm == SchnorrAlgorithm::Ed25519 {
+                let key_id = match key_id.name.as_ref() {
+                    "key_1" => ic_ed25519::MasterPublicKeyId::Key1,
+                    "test_key_1" => ic_ed25519::MasterPublicKeyId::TestKey1,
+                    _ => return Err(Error::UnknownKeyIdentifier),
+                };
+
+                let mk = ic_ed25519::PublicKey::mainnet_key(key_id);
+                let inner = MasterPublicKeyInner::Ed25519(mk);
+                return Ok(Self { inner });
+            }
+        }
+
+        let _ignored = key_id;
+        Err(Error::AlgorithmNotSupported)
+    }
+}
+
+impl TryFrom<&VetKDKeyId> for MasterPublicKey {
+    type Error = Error;
+
+    fn try_from(key_id: &VetKDKeyId) -> Result<Self, Self::Error> {
+        #[cfg(feature = "vetkeys")]
+        {
+            if let Some(mk) = ic_vetkeys::MasterPublicKey::for_mainnet_key(key_id) {
+                let inner = MasterPublicKeyInner::VetKD(mk);
+                return Ok(Self { inner });
+            }
+        }
+
+        let _ignore = key_id;
+        Err(Error::AlgorithmNotSupported)
+    }
+}
+
+enum DerivedPublicKeyInner {
+    #[cfg(feature = "secp256k1")]
+    EcdsaSecp256k1((ic_secp256k1::PublicKey, [u8; 32])),
+    #[cfg(feature = "secp256k1")]
+    Bip340Secp256k1((ic_secp256k1::PublicKey, [u8; 32])),
+    #[cfg(feature = "ed25519")]
+    Ed25519((ic_ed25519::PublicKey, [u8; 32])),
+    #[cfg(feature = "vetkeys")]
+    VetKD(ic_vetkeys::DerivedPublicKey),
+}
+
+/// The canister's master public key of a threshold signature system
+///
+/// Each canister gets its own canister master key, which is derived from
+/// the system master key.
+pub struct CanisterMasterKey {
+    inner: DerivedPublicKeyInner,
+}
+
+impl CanisterMasterKey {
+    /// Derive the public key from a canister key and a single contextual input
+    ///
+    /// This is the only supported method for VetKD keys
```

Comment:
It might make sense to add a sentence saying that [if the method is invoked with an empty derivation context for vetKD, it simply returns the canister master key](https://github.com/dfinity/vetkeys/blob/83b887f220a2c1c40713a3512ce5a9994d5ec4c6/backend/rs/ic_vetkeys/src/utils/mod.rs#L454-L456).

Or should we go as far as even disallowing empty contexts here?

Reply:
IMO this should replicate the management canister behavior as closely as possible, so I'd lean to documenting the behavior since it *is* potentially quite surprising.